### ISSUE TITLE
Ability to start browsers in private / incognito mode

### ISSUE THREAD

**Issue by johnste** [johnste/finicky#49]

**Comment by niobos** [comment]
An alternative would be to allow an arbitrary script to be run (with the URL as parameter). The script can then launch whatever browser with whatever options it wants

**Comment by wenisy** [comment]
any new updates about it?

**Comment by johnste** [comment]
Sorry, no real updates. I am however looking into this currently (as well as #40) and trying to figure out a good approach for it. I've tried a implementation before but it didn't work well unfortunately.

**Comment by johnste** [comment]
_Ok, so sometimes inspiration (and available time slots) just strikes I guess._

I've just published an ALPHA release of basic Chrome Incognito support, just to try a few things out. Please check  it out and tell me if it works for you! It's a really early version and the feature might not work as expected, which is why feedback is most welcome.

https://github.com/johnste/finicky/releases/tag/v2.3-alpha

**Comment by johnste** [comment]
I'm interested to hear from people who've tried the 2.3 alpha if it's been working well for them. There's no metrics collection in finicky so I rely on people giving feedback here or over on [Twitter](https://twitter.com/johnste_) to know if stuff is working.

**Comment by bryanhuntesl** [comment]
@johnste - I've got an interesting behaviour - for some reason finicky is opening URL's in Google Chrome in incognito mode - could this be related? Here's my config, it's pretty basic.

```
module.exports = {
        defaultBrowser: "Firefox",
        handlers: [
            {
              match: finicky.matchHostnames(["docs.google.com", "drive.google.com", "youtube.com", "facebook.com", "global.gotomeeting.com"]),
                browser: "Google Chrome"
            },
            {
                match: finicky.matchHostnames(["apple.com"]),
                browser: "Safari"
            }
        ]
};
```

**Comment by bryanhuntesl** [comment]
Locally built from master - it doesn't exhibit the same behaviour - https://docs.google.com is opened in a standard Chrome window.

> @johnste - I've got an interesting behaviour - for some reason finicky is opening URL's in Google Chrome in incognito mode - could this be related? Here's my config, it's pretty basic.
> 
> ```
> module.exports = {
>         defaultBrowser: "Firefox",
>         handlers: [
>             {
>               match: finicky.matchHostnames(["docs.google.com", "drive.google.com", "youtube.com", "facebook.com", "global.gotomeeting.com"]),
>                 browser: "Google Chrome"
>             },
>             {
>                 match: finicky.matchHostnames(["apple.com"]),
>                 browser: "Safari"
>             }
>         ]
> };
> ```

**Comment by bucweat** [comment]
Noticed that 2.4 includes "You can now start browsers by file path: e.g. /Applications/Slack.app." Would it be possible to allow for a script (this is mentioned in at least one other feature request #89)? Something that would allow you to provide something like:

```
/Applications/Microsoft\ Edge.app/Contents/MacOS/Microsoft\ Edge --inprivate {url}

or

open /Applications/PrivateEdge.app --args {url} 
```
PrivateEdge.app is a application wrapper (e.g. folder structure with MacOS and Contents subfolders) containing a shell script with the first line above. In this case it would open link in Edge in private mode.

Just an idea. 

I did try making a "PrivateEdge.app" 
```
PrivateEdge.app
├── Contents
│   └── MacOS
│       └── PrivateEdge
└── PkgInfo
```
where PrivateEdge contained 
```
/Applications/Microsoft\ Edge.app/Contents/MacOS/Microsoft\ Edge --inprivate $1
``` 
This worked fine on command line and I was able to set as browser for some websites in finicky config, but for some reason it opens a normal Edge window instead of a private window. My default browser is Safari BTW.

**Comment by meiser79** [comment]
I'm also interested in private/incognito mode, but more in a generic way to e.g. also support Opera.

**Comment by johnste** [comment]
Yeah there's an issue for it. Historicalöy I've had issues reliably
starting private/incognito browsers so that's why its still not ready.

**Comment by meiser79** [comment]
Thanks to the example in #153, I was able to find the right config to open Opera in private mode.

```
   handlers: [
      {
         // Open links in Opera when the option key is pressed
         // Valid keys are: shift, option, command, control, capsLock, and function.
         // Please note that control usually opens a tooltip menu instead of visiting a link
         match: () => finicky.getKeys().option,
         browser: ({ urlString }) => ({
            name: "Opera",
            args: ["-private", urlString],
         })
      },
   ]
```
It's not yet perfect as new links are opened in a new private windows instead of a new private tab if Opera is already started.

**Comment by johnste** [comment]
Yes, the reason it's not officially supported is because of issues like the one you mentioned.

**Comment by tonymet** [comment]
`args: ["--incognito"]` works for Chrome
```
   // PERSONAL PROFILE
      match: [
        "mail.google.com/*", // match google.com urls      ],
      browser: {
          name:  "Google Chrome",
          profile: "Profile 1", 
          args: ["--incognito"],
      },
    }
  ]
```

**Comment by residentsummer** [comment]
This works for me quite reliably on `v4.0.0-beta.3`:

```
    {
      match: "python.org/*",
      browser: (url, options) => ({
          name: "Brave Browser",
          args: ["-n", "--args", "--incognito", "--profile-directory=Profile 3",  url.href]
      }),
    },
```

The only issue that is left is a peculiar one: if target browser window is on the another display, it won't activate. Instead, the last active window on the current display will. This behaviour is unrelated to incognito mode, though. I believe it's chromium issue or quirk of Apple's window management.
